Add the ability to pin Issues

JakobDev opened the pull request:
This adds the ability to pin important Issues and Pull Requests. You can also move pinned Issues around to chnage their Position. Resolves #2175.

**Screenshots:**
![grafik](https://user-images.githubusercontent.com/15185051/235123207-0aa39869-bb48-45c3-abe2-ba1e836046ec.png)
![grafik](https://user-images.githubusercontent.com/15185051/235123297-152a16ea-a857-451d-9a42-61f2cd54dd75.png)

The Design was mostly copied from the Projects Board.

**Implementation:**
This uses a new `pin` Column in the `issue` table.  If the value is set to 0, the Issue is not pinned. If it's set to a bigger value, the value is the Position. 1 means it's the first pinned Issue, 2 means it's the second one etc. This is dived into Issues and Pull requests for each Repo.

**TODO:**
You can currently pin as many Issues as you want. Maybe we should add a Limit, which is configurable. GitHub uses 3, but I prefer 6, as this is better for bigger Projects, but I'm open for suggestions.

I also need to add a Migration and Tests before this PR is ready to merge.

**The feature itself is currently fully working, so tester who may find weird edge cases are very welcome!**

wxiaoguang commented on models/issues/issue.go:
```diff
@@ -127,6 +127,7 @@ type Issue struct {
 	PullRequest      *PullRequest     `xorm:"-"`
 	NumComments      int
 	Ref              string
+	Pin              int `xorm:"DEFAULT 0"`
```
Could it be a two-word column name, eg: `PinOrder`?

`Pin` reads like "pinned" or not.

silverwind commented:
<img width="237" alt="image" src="https://user-images.githubusercontent.com/115237/235150164-b699cf50-68da-4c9c-8ed7-8ec3a0f0d8e2.png">

Border is too thick, please use 1px border.

silverwind commented:
> GitHub uses 3, but I prefer 6, as this is better for bigger Projects, but I'm open for suggestions.

I'd say 3 is fine as two rows of pins may be obnoxious.

Is a "X" icon implemented in the boxes where users with sufficient permission can un-pin?

JakobDev commented:
> Border is too thick, please use 1px border.

Done

> I'd say 3 is fine as two rows of pins may be obnoxious.

I still think 6 would be better for bigger projects, but let's see what the other says

> Is a "X" icon implemented in the boxes where users with sufficient permission can un-pin?

This is now added. I have also found a Bug in Tippy: If a Element with a Tooltip is removed, while the Tooltip is shown, the Tooltip is still shown.

![grafik](https://user-images.githubusercontent.com/15185051/235168825-eba34c6c-d520-40f6-8ad3-84c168f0155f.png)

silverwind commented:
> This is now added. I have also found a Bug in Tippy: If a Element with a Tooltip is removed, while the Tooltip is shown, the Tooltip is still shown.

I think it's a case we hadn't seen yet. You can probably just `el._tippy.destroy()` before `el.remove()`.

silverwind commented:
Are you on Windows? I think we have another [newline issue in generated file](https://drone.gitea.io/go-gitea/gitea/73966/1/9).

JakobDev commented:
> Are you on Windows?

My main Computer with Manjaro has some Problems at the Moment, so I'm currently forced to use this stupid OS on another Computer. I don't even know which script touches this file at all. Maybe you should just ignore if there is a final newline, if that makes problem. A final newline is not very important on such a large autogenrated file.

JakobDev commented:
> You can probably just [el._tippy.destroy()](https://atomiks.github.io/tippyjs/v5/methods/#destroy) before [el.remove()](https://developer.mozilla.org/en-US/docs/Web/API/Element/remove).

This looks like a private function. Should this really be used?

silverwind commented:
> > You can probably just [el._tippy.destroy()](https://atomiks.github.io/tippyjs/v5/methods/#destroy) before [el.remove()](https://developer.mozilla.org/en-US/docs/Web/API/Element/remove).
> 
> This looks like a private function. Should this really be used?

If you can obtain the reference to the tippy instance through other means, go for it. `_tippy` is only for cases where it's not possible to get the instance otherwise.

silverwind commented:
> > Are you on Windows?
> 
> My main Computer with Manjaro has some Problems at the Moment, so I'm currently forced to use this stupid OS on another Computer. I don't even know which script touches this file at all. Maybe you should just ignore if there is a final newline, if that makes problem. A final newline is not very important on such a large autogenrated file.

So at least that confirms it's a Windows issue. File is generated during `make tidy`. We can add normalization for it, but it appears that normalization hadn't worked before in the swagger case via `sed`, so I guess we can do it in the golang script for this case.

silverwind commented:
https://github.com/go-gitea/gitea/pull/24407 should fix the newline issue.

delvh commented:
What I've noticed already:
Yes, it's one possible approach to store the pin order directly inside the issues.
However, this approach also makes problems:
- PRs also need that attribute as you can see at the moment. But what is a pinned PR?
- This is often unneeded as it's pretty rare that you need to display the pinned issues (basically only on the "view issues" page)
- What's the behavior for when the issue isn't pinned? `PinnedOrder = 0`?
- You will only ever be able to pin your own issues and foreign issues from other repos will never be possible (not sure if this is actually useful or not)

All of these problems can be solved by using the following approach:
Instead of storing for each issue the ranking it is pinned in, let's store an external
```go
type PinnedIssue {
  repoID int64 UNIQUE(issue_repo) // the repo pinning this issue, not necessarily the one owning it. Additionally, we can even think about whether this is the primary key for this struct, as every repo should likely only have one set of pinned issues…
  issueID int64 UNIQUE(issue_repo)
  order int64 // or a smaller type, it's not going to be bigger than 1000…
  createdUnix timestamp
}
```
Then you can query the pinned issues of a repository with the query
```go
e.Where(PinnedIssue{repoID: repoID}).Asc("order, createdUnix").Find(<result list>)
```
Which should suffice for all cases where you actually want to separate the pinned issues.
Perhaps you'll need the second query
```go
e.Get(PinnedIssue{repoID: <repoID>, issueID: <issueID>}).Exists()
```
to find out if a repo pins a single issue (It looked like you wanted to display a special icon on the issue view/edit page depending on if it is pinned or not when I skimmed through the PR changes).

What do you think?

(Every code I wrote here should be treated as pseudo-working-code. It may work, but it probably needs some adjustments.)

JakobDev commented:
> PRs also need that attribute as you can see at the moment. But what is a pinned PR?

If you are on the Issues page, you see pinned Issues, if you are on the PR page, you see the pinned PRs as shown the Screenshot.

> This is often unneeded as it's pretty rare that you need to display the pinned issues (basically only on the "view issues" page)

Sure, you only need this foe the View Issues page and the Sidebar for a single Issue, but why is that a Problem?

> What's the behavior for when the issue isn't pinned?` PinnedOrder = 0?`

Then it's simply not pinned. I'm not sure what you mean.

> You will only ever be able to pin your own issues and foreign issues from other repos will never be possible (not sure if this is actually useful or not)

This is part of the Design. I think it should not be possible to pin Issues from other Repos. This would only be bring a benefit in a handful of situations, but it will make the Implementation more complex. It also confuses Users who see Issues from another Repo on the Issue page of a Repo.

silverwind commented on options/locale/locale_en-US.ini:
```diff
@@ -1463,6 +1463,9 @@ issues.attachment.open_tab = `Click to see "%s" in a new tab`
 issues.subscribe = Subscribe
 issues.unsubscribe = Unsubscribe
+issues.pin = Pin
+issues.unpin = Unpin
```
Move these to to top-level please, outside any section, e.g. translation keys `pin` and `unpin`.

JakobDev replied:
You mean `pin` instead of `repo.issues.pin`? Why? They belong to the Issues of a Repo.

silverwind replied:
Yes, but these terms are universal and usable elsewhere, like "pin" for project card, etc. We need to cut down on translating the same thing 5 times, unless they mean different things in different contexts, of course.

silverwind commented on templates/repo/issue/list.tmpl:
```diff
@@ -2,6 +2,88 @@
+									{{else}}
+										{{svg "octicon-issue-opened" 16 "text green"}}
+									{{end}}
+								{{end}}
```
I think there is a shared template for these icons somewhere, if not, make one where possible.

JakobDev replied:
Why do we need a template for the octicon? It's just one line.

silverwind replied:
I'm talking about the whole block above, which is like 30 lines and is definetly something that should not be repeated across the codebase.

silverwind replied:
The code is exactly the same as in `templates/shared/issuelist.tmpl`, so I suggest you create `templates/shared/issueicon.tmpl`, move the code there and then import is via `{{template}}` in both places.

silverwind commented on templates/repo/issue/list.tmpl:
```diff
@@ -2,6 +2,88 @@
 	<div class="ui container">
+
+	{{if .PinnedIssues}}
+		<div id="issue-pin-div" {{if .IsRepoAdmin}}data-is-repo-admin{{end}} style="display: flex; flex-grow: 1; flex-wrap: wrap;">
```
No inline styles please.

silverwind commented on templates/repo/issue/list.tmpl:
```diff
@@ -2,6 +2,88 @@
+									{{end}}
+								{{end}}
+							</span>
+							<a class="project-board-title gt-vm" href="{{.Link}}">
```
`project-board-title`?

JakobDev replied:
Yes, this a a copy/paste error. I will copy it under a new name.

silverwind commented on templates/repo/issue/list.tmpl:
```diff
@@ -25,6 +107,7 @@
 				{{end}}
 			{{end}}
 		</div>
+
```
```suggestion
```

silverwind commented on web_src/js/features/repo-issue-list.js:
```diff
@@ -119,8 +120,80 @@ function initRepoIssueListAuthorDropdown() {
 
+function initPinRemoveButton() {
+  for (const button of document.getElementsByName('issue-unpin-button')) {
+    button.addEventListener('click', (event) => {
+      const elem = event.currentTarget;
+      const id = Number(elem.getAttribute('data-issue-id'));
+
+      // Send the unpin request
+      $.ajax({
+        url: elem.getAttribute('data-unpin-url'),
```
```suggestion
    button.addEventListener('click', (e) => {
      const el = e.currentTarget;
      const id = Number(el.getAttribute('data-issue-id'));

      // Send the unpin request
      $.ajax({
        url: el.getAttribute('data-unpin-url'),
```

silverwind commented on web_src/js/features/repo-issue-list.js:
```diff
@@ -119,8 +120,80 @@ function initRepoIssueListAuthorDropdown() {
+  }
+}
+
+function pinMoveEnd(event) {
+  const url = event.item.getAttribute('data-move-url');
+  const id = Number(event.item.getAttribute('data-issue-id'));
+  $.ajax({
```
```suggestion
function pinMoveEnd(e) {
  const url = e.item.getAttribute('data-move-url');
  const id = Number(e.item.getAttribute('data-issue-id'));
  $.ajax({
```

silverwind commented on web_src/js/features/repo-issue-list.js:
```diff
@@ -119,8 +120,80 @@ function initRepoIssueListAuthorDropdown() {
+    error: (xhr) => {
+      throw new Error(xhr.responseText);
+    }
+  });
```
Could you use `fetch` please? There are existing examples in the codebase already.

silverwind commented on web_src/js/features/repo-issue-list.js:
```diff
@@ -119,8 +120,80 @@ function initRepoIssueListAuthorDropdown() {
+
+function initIssuePinSort() {
+  const pinDiv = document.getElementById('issue-pin-div');
+
+  if (pinDiv === null) {
+    return;
+  }
```
```suggestion
  if (!pinDiv) return;
```

silverwind commented on web_src/js/features/repo-issue-list.js:
```diff
@@ -119,8 +120,80 @@ function initRepoIssueListAuthorDropdown() {
+  }
+
+  // If the User is not a Repo Admin, we don't need to proceed
+  if (!pinDiv.hasAttribute('data-is-repo-admin')) {
+    return;
+  }
```
```suggestion
  if (!pinDiv.hasAttribute('data-is-repo-admin')) return;
```

silverwind commented on web_src/js/features/repo-issue-list.js:
```diff
@@ -119,8 +120,80 @@ function initRepoIssueListAuthorDropdown() {
+  initPinRemoveButton();
+
+  // If only one issue pinned, we don't need to make this Sortable
+  if (pinDiv.children.length < 2) {
+    return;
+  }
```
```suggestion
  if (pinDiv.children.length < 2) return;
```

silverwind commented on web_src/js/features/repo-issue-list.js:
```diff
@@ -119,8 +120,80 @@ function initRepoIssueListAuthorDropdown() {
+    group: 'shared',
+    animation: 150,
+    ghostClass: 'card-ghost',
+    onEnd: pinMoveEnd
```
```suggestion
    onEnd: pinMoveEnd,
```

silverwind commented on templates/repo/issue/list.tmpl:
```diff
@@ -2,6 +2,88 @@
+					{{if or .Labels .Assignees}}
+					<div class="extra content labels-list gt-p-0 gt-pt-2">
+						{{range .Labels}}
+							<a target="_blank" href="{{$.RepoLink}}/issues?labels={{.ID}}">{{RenderLabel $.Context .}}</a>
```
We don't usually open stuff in new tabs, what is the reason for it here?

silverwind commented on templates/repo/issue/list.tmpl:
```diff
@@ -2,6 +2,88 @@
+						{{end}}
+						<div class="right floated">
+							{{range .Assignees}}
+								<a target="_blank" href="{{.HomeLink}}" data-tooltip-content="{{$.locale.Tr "repo.projects.column.assigned_to"}} {{.Name}}">{{avatar $.Context . 28 "mini gt-mr-3"}}</a>
```
Same

silverwind commented on templates/repo/issue/list.tmpl:
```diff
@@ -2,6 +2,88 @@
+								{{.Title}}
+							</a>
+							{{if $.IsRepoAdmin}}
+								<span name="issue-unpin-button" style="float: right;" data-tooltip-content={{$.locale.Tr "repo.issues.unpin_issue"}} data-issue-id="{{.ID}}" data-unpin-url="{{$.Link}}/unpin/{{.ID}}">
```
No inline styles please

JakobDev commented:
I had now added a configurable limit for the pinned Issues

JakobDev commented:
I have now added to pin and unpin Events to the History. this was way easier and faster than I thought. This PR is now feature complete. The only remaining things are the Tests and the Migration.
![grafik](https://user-images.githubusercontent.com/15185051/235640736-cdb5ff8b-f3ee-4827-9798-400e66330664.png)

JakobDev commented:
I have now added tests and the migration. This PR is now ready to review!

wxiaoguang commented on custom/conf/app.example.ini:
```diff
@@ -1070,6 +1070,8 @@ ROUTER = console
 ;; List of reasons why a Pull Request or Issue can be locked
 ;LOCK_REASONS = Too heated,Off-topic,Resolved,Spam
+;; Maximum number of pinned Issues
+;MaxPinned = 6
```
That's not the style of other config options 😀

JakobDev replied:
Fixed

wxiaoguang commented on web_src/js/features/repo-issue-list.js:
```diff
@@ -119,8 +120,69 @@ function initRepoIssueListAuthorDropdown() {
+      'Content-Type': 'application/json',
+    },
+  });
+  initPinRemoveButton();
```
duplicate call to initPinRemoveButton ? some element may have duplicate `click` events?

JakobDev replied:
I initially though the Icon would lose the Event when the div is moved, but is appears this is not true, so I removed it now

JakobDev commented on web_src/js/features/repo-issue-list.js:
```diff
@@ -119,8 +120,69 @@ function initRepoIssueListAuthorDropdown() {
+          // Delete the tooltip
+          el._tippy.destroy();
+          // Remove the Card
+          document.querySelector(`div[data-issue-id="${id}"]`).remove();
```
`el` is just the X Icon, but we need to remove the full div

wxiaoguang reviewed (commented):
Overall looks good, some nits in my mind:

* some CSS nits, eg: "float right" seems not good, "pinned-issue-card" max-width is complex?
* use `await` for fetch instead of "then" ?
* not sure about whether some CSS classes are stable enough (card/meta/extra/right floated)

ps: I am quite busy recently (on some refactoring PRs), so I haven't gotten enough time to do a full review/test. And I have no writer permission so I can't help to improve these nits.

So if some reviewers could review and approve, it would help.

JakobDev commented:
> some CSS nits, eg: "float right" seems not good, "pinned-issue-card" max-width is complex?

I'm not very good at CSS. What's the Problem here with` float: right`?

> use `await` for fetch instead of "then" ?

Done

> not sure about whether some CSS classes are stable enough (card/meta/extra/right floated)

So I should copy them as a new Class?

silverwind commented:
> I'm not very good at CSS. What's the Problem here with float: right?

Float is meant for "floating" a elements besides a text, not for general UI layout like we do in some places. For a horizontal flexbox pin box header, e.g. title and "x" icon, you can do something like:

```handlebars
<div class="gt-df gt-ac">
  <div class="gt-df gt-ac gt-f1">
    <div class="gt-ellipsis">{{title}}</div>
  </div>
  <div class="gt-df gt-ac">{{svg}}</div>
</div>
```

JakobDev commented:
I have now updated the Layout with the Code from silverwind

silverwind commented:
Issue sidebar button margins fixed:

<img width="298" alt="Screenshot 2023-05-09 at 22 40 21" src="https://github.com/go-gitea/gitea/assets/115237/82a58126-3e65-4f34-bea5-052da856d426">

silverwind commented:
Doing some more style fixes...

silverwind commented:
Pushed some style fixes, looks much better now and uses CSS `grid`:

<img width="1237" alt="Screenshot 2023-05-09 at 23 19 15" src="https://github.com/go-gitea/gitea/assets/115237/9e75bb31-37d4-4e08-944a-822486470f7d">
<img width="1236" alt="Screenshot 2023-05-09 at 23 19 27" src="https://github.com/go-gitea/gitea/assets/115237/520cb410-9bde-4e57-9ec0-a7d45b7b6f74">

wxiaoguang commented on routers/web/repo/issue_pin.go:
```diff
@@ -0,0 +1,83 @@
+		return
+	}
+
+	err = issue.PinOrUnpin(ctx, ctx.ContextUser)
```
It seems not right. ContextUser is not Doer

JakobDev replied:
Done

silverwind commented on templates/shared/issueicon.tmpl:
```diff
@@ -0,0 +1,18 @@
+
```
```suggestion
```

silverwind commented on templates/repo/issue/view_content/comments.tmpl:
```diff
@@ -11,7 +11,8 @@
 		32 = DISMISSED_REVIEW, 33 = COMMENT_TYPE_CHANGE_ISSUE_REF, 34 = PR_SCHEDULE_TO_AUTO_MERGE,
-		35 = CANCEL_SCHEDULED_AUTO_MERGE_PR -->
+		35 = CANCEL_SCHEDULED_AUTO_MERGE_PR,
+		36 = Pin, 37 = Unpin -->
```
Should these be uppercase and on same line as above?

JakobDev commented:
@silverwind Thanks for helping with the frontend stuff! I really appreciate that!

When #24669 is merged, I can also use the new Unpin Icon in this PR

lunny commented:
> @silverwind Thanks for helping with the frontend stuff! I really appreciate that!
> 
> When #24669 is merged, I can also use the new Unpin Icon in this PR

Merged.

JakobDev commented:
I updated the PR so it now uses the new Icon

silverwind commented:
@JakobDev can you fix the issue title rendering in pinned issues like done [here](https://github.com/go-gitea/gitea/pull/24639/files#diff-40b7988fcd35c50d12157e06dc96b586404a4d1d3772a4b257340f11295bec5aR64) so that content inside backticks renders as code block:

<img width="423" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/bb118a66-9dfd-4837-894a-d6930c8a4a13">

silverwind commented:
Also, one merge conflict in `options/locale/locale_en-US.ini` appeared.

lunny commented:
unpin has no response and it returned 500 in fact.
``
2023/05/13 14:31:43 ...s/process/manager.go:207:remove() [T] [645f2ecf] Done 645f2ecf: DELETE: /actions/release-action/issues/unpin/203018
2023/05/13 14:31:43 [645f2ecf] router: completed DELETE /actions/release-action/issues/unpin/203018 for 127.0.0.1:49514, 500 Internal Server Error in 115.2ms @ repo/issue_pin.go:35(repo.IssueUnpin)
```

JakobDev commented:
> @JakobDev can you fix the issue title rendering in pinned issues like done [here](https://github.com/go-gitea/gitea/pull/24639/files#diff-40b7988fcd35c50d12157e06dc96b586404a4d1d3772a4b257340f11295bec5aR64) so that content inside backticks renders as code block:

Done

> Also, one merge conflict in `options/locale/locale_en-US.ini` appeared.

Done

> unpin has no response

This is expected. This is only called by a JS Function and not visited by the User directly.

> and it returned 500 in fact.

I can't reproduce this on my Dev System. Can you find out what exactly causes this error, so I can fix it?

I had also added, that when `MAX_PINNED` is set to 0 in the Config, the button to Pin or Unpin a Issue is not shown.

wxiaoguang commented on routers/web/repo/issue_pin.go:
```diff
@@ -0,0 +1,83 @@
+func IssueUnpin(ctx *context.Context) {
+	issue, err := issues_model.GetIssueByIndex(ctx.Repo.Repository.ID, ctx.ParamsInt64(":id"))
+	if err != nil {
+		ctx.JSON(http.StatusInternalServerError, err.Error())
```
Have we ever exposed the internal `err` to end users?

In most causes, they are not IIRC. People worried about that these errors might contain sensitive (server) messages?

wxiaoguang commented on models/issues/issue.go:
```diff
@@ -2507,3 +2509,178 @@ func DeleteOrphanedIssues(ctx context.Context) error {
+		return err
+	}
+
+	_, err = db.GetEngine(db.DefaultContext).Table("issue").
```
To be strict, database operations should in one context related translation. 

I am not sure it is acceptable to use more `db.DefaultContext` in new code.

JakobDev replied:
Why is `db.DefaultContext` a problem?

wxiaoguang replied:
1. no transaction on db.DefaultContext
2. if a new transaction is started by db.DefaultContext here but not the web's context, it might cause conflicts with web transaction, eg: can not see other's updates if the web transaction is not committed.

a1012112796 commented on custom/conf/app.example.ini:
```diff
@@ -1067,6 +1067,9 @@ ROUTER = console
 ;LOCK_REASONS = Too heated,Off-topic,Resolved,Spam
+;; Maximum number of pinned Issues
+;; Set to 0 to disable pinning Issues
+;MAX_PINNED = 6
```
wonder why limit max number?

eeyrjmr replied:
> wonder why limit max number?

At a guess UX concerns. 3 or 6 is a reasonable number (in rows of 3), anymore and humans will lose focus. 
Take this to the extreme and be able to pin inf... Then the point of pinning is lost while also inheriting a major UX issue

a1012112796 commented on models/issues/issue.go:
```diff
@@ -2507,3 +2509,178 @@ func DeleteOrphanedIssues(ctx context.Context) error {
+	}
+
+	var maxPin int
+	_, err := db.GetEngine(ctx).SQL("SELECT MAX(pin_order) FROM issue WHERE repo_id = ? AND is_pull = ?", issue.RepoID, issue.IsPull).Get(&maxPin)
```
I think it's necessary to use TxContext in this function. usage example:
https://github.com/go-gitea/gitea/blob/20356cfd8ee79de898243e59ef6ca63cd3516c64/models/issues/issue.go#L502-L506

JakobDev replied:
It currently works, so what's the difference?

silverwind commented on modules/setting/repository.go:
```diff
@@ -227,8 +228,10 @@ var (
+			MaxPinned   int
 		}{
 			LockReasons: strings.Split("Too heated,Off-topic,Spam,Resolved", ","),
+			MaxPinned:   6,
```
```suggestion
			MaxPinned:   3,
```

As discussed, I think we better match this default to GitHub. No strong opinion, but two rows already seems too much to me.

silverwind commented on custom/conf/app.example.ini:
```diff
@@ -1025,6 +1025,9 @@ LEVEL = Info
 ;LOCK_REASONS = Too heated,Off-topic,Resolved,Spam
+;; Maximum number of pinned Issues
+;; Set to 0 to disable pinning Issues
+;MAX_PINNED = 6
```
```suggestion
;MAX_PINNED = 3
```

Also add an entry to https://github.com/go-gitea/gitea/blob/main/docs/content/doc/administration/config-cheat-sheet.en-us.md please. Every config addition should be both in `app.example.ini` and there.

JakobDev commented:
I'm still more in favour of 6 as default, as I think it's better for larger Projects. It also brings Room for different kind of Issues e.g. first Row are current discussions, second row are currently critical Bug reports. To solve this discussion, I think it's the best to make a little vote, to see what more people like.

To vote, add a reaction to this comment:
Heart: 3 as default value
Rocket: 6 as default value

I will use the one that has more votes as default value.

JakobDev commented:
Does anybody knew why `_, err = db.GetEngine(dbctx).Table("issue").` raises a `ineffectual assignment to err (ineffassign)` error?

delvh commented:
What that says is you assign `err` without using it.

eeyrjmr commented:
> I'm still more in favour of 6 as default,
So since the MAX_PIN is in app.ini, what is to stop me putting 999999 in that field ?
Is there a hard-coded max ?

JakobDev commented:
> What that says is you assign `err` without using it.

Thanks.

Maybe it would be better if we call this "err is assigned but not used" or something else to make this clear.

> So since the MAX_PIN is in app.ini, what is to stop me putting 999999 in that field ?
> Is there a hard-coded max ?

Nothing, but you need to be the Server Admin to do this. There are Users who use a hosted Gitea (e..g. I use Codeberg for my Projects), what means they can't change this themself, so I think we should use a good default value here that is capable for most Users. There is also nothing that forces you to use the max Number of Pinned Issues. If you like 3 more than 6, you still can only Pin 3 Issues while 6 is allowed.

And I also think this will not be changed by most Admins. They will most likely just use the default value and didn't take a look, if this can be configured.

wxiaoguang commented on models/migrations/migrations.go:
```diff
@@ -491,8 +491,12 @@ var migrations = []Migration{
+	NewMigration("Add PinOrder Column", v1_20.AddPinOrderToIssue),
 	// v257 -> v258
 	NewMigration("Add Actions Artifact table", v1_20.CreateActionArtifactTable),
+	// v258 -> 259
+	NewMigration("Add PinOrder Column", v1_20.AddPinOrderToIssue),
```
Oops, the annoying Gitea's migration framework ..... 😭 

Otherwise, overall LGTM.

JakobDev replied:
What's wrong with this line?

wxiaoguang replied:
Merge problem.

![image](https://github.com/go-gitea/gitea/assets/2114189/ba50b14e-bbef-4ccb-9b37-0c5b6a19264d)

JakobDev commented:
Most people has voted for 3, so I changed it now to 3. I also added Documentation in `docs/content/doc/administration/config-cheat-sheet.en-us.md`.

Does it make sense to run `UPDATE issue SET pin_order = 0 WHERE pin_order > MAX_PINNED` on every startup to make sure, that not more Issues are pinned than allowed after a configuration change or can we ignore this?

jolheiser commented:
I agree with @wxiaoguang, though perhaps a doctor command could be introduced for it, or a task on the admin page. 

It may be nice for admins to be able to do it without sticking fingers in the DB, but I don't think it requires a startup query.

JakobDev commented:
I think adding a new Button or a new doctor command just for the rare case somebody changed a specific value in the config is a little bit overkill, so I keep things as they are

silverwind commented on templates/repo/issue/view_content/comments.tmpl:
```diff
@@ -11,7 +11,7 @@
 		29 = PULL_PUSH_EVENT, 30 = PROJECT_CHANGED, 31 = PROJECT_BOARD_CHANGED
 		32 = DISMISSED_REVIEW, 33 = COMMENT_TYPE_CHANGE_ISSUE_REF, 34 = PR_SCHEDULE_TO_AUTO_MERGE,
-		35 = CANCEL_SCHEDULED_AUTO_MERGE_PR -->
+		35 = CANCEL_SCHEDULED_AUTO_MERGE_PR, 36 = PIN, 37 = UNPIN -->
```
```suggestion
		35 = CANCEL_SCHEDULED_AUTO_MERGE_PR, 36 = PIN_ISSUE, 37 = UNPIN_ISSUE -->
```

Think this is a better label. These strings don't mean anything right, they are not actually used anywhere besides this comment?

silverwind replied:
Ok, thanks

silverwind commented:
Test failure is related:

```
    api_issue_pin_test.go:170: 
        	Error Trace:	/home/runner/work/gitea/gitea/tests/integration/api_issue_pin_test.go:170
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 0
        	Test:       	TestAPIListPinnedIssues
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0
```

Once fixed, I think we can merge.

JakobDev commented:
Is the test failure maybe just a Bug in the test suite? It only fails for Postgres, but not for the other Databases. And I only use basic SQL, so the Database should not be a Problem.

JakobDev commented:
The tests were successful this time, so this was a Bug

JakobDev commented:
This looks more like a Bug in the test suite. Correct me if I'm wrong, but I think the tests of a test suite should be always reproduceable (things like calling a external API are exceptions, but this does not happen here). The tests are executed in a clean Environment, so they should always behave the same. To be clear, I'm not saying that it is impossible that this Code does have a Bug that is only triggered under weird reproduceable, but I'm saying that the Tests should not fail one time and run sucessfully the second time without changing any Code.

Anyway, I'm updating the Branch to retrigger the Tests to see if it will run.

silverwind commented:
We now have this same [issueicon.tmpl](https://github.com/go-gitea/gitea/blob/main/templates/shared/issueicon.tmpl) on the main branch which should be compatible with the usage here, so the file can be removed here.

JakobDev commented:
Wait a moment

silverwind commented:
Okay, tell me when it's ready.

JakobDev commented:
It is now ready. I have found a outdated TODO comment that I had removed now to not confuse others.

delvh commented on models/issues/issue.go:
```diff
@@ -684,3 +686,180 @@ func (issue *Issue) GetExternalID() int64 { return issue.OriginalAuthorID }
+// IsNewPinnedAllowed returns if a new Issue or Pull request can be pinned
+func IsNewPinAllowed(ctx context.Context, repoID int64, isPull bool) (bool, error) {
+	var maxPin int
+	_, err := db.GetEngine(ctx).SQL("SELECT MAX(pin_order) FROM issue WHERE repo_id = ? AND is_pull = ?", repoID, isPull).Get(&maxPin)
```
Use `COUNT` instead of `MAX`.
That will fail less often (i.e. if there are spaces between pin orders).

delvh commented on web_src/css/repo.css:
```diff
@@ -3387,3 +3387,37 @@ tbody.commit-list {
+  background: var(--color-card);
+}
+
+.pinned-issue-card .meta a {
+  color: inherit;
+}
+
+.pinned-issue-card .meta a:hover {
```
Why `.meta`?
Why not simply omit this intermediary selector?

silverwind commented:
Noticed one bug: I can not unpin issues from the issue page because the `ID` value is incorrect, so it sends a `DELETE` request, gets 204, the issue is deleted in UI, but comes back on next page reload.

@JakobDev maybe you can followup with a fix PR that adresses above comments too.

lunny commented:
> Noticed one bug: I can not unpin issues from the issues list page because the `ID` value is incorrect, so it sends a `DELETE` request, gets 204, the box is removed in UI, but comes back on next page reload. Unpin from single issue page works fine.
> 
> @JakobDev maybe you can followup with a fix PR that adresses above comments too.

I think I have reported that I cannot remove the pinned issue?

JakobDev commented:
> Noticed one bug: I can not unpin issues from the issues list page because the `ID` value is incorrect, so it sends a` DELETE `request, gets 204, the box is removed in UI, but comes back on next page reload. Unpin from single issue page works fine.

I'm sorry, but I can reproduce this on a brand new setup on Firefox and Chrome. I can unpin Issues using the X without problems. Can you maybe investigate what causes the wrong ID in your case?

> maybe you can followup with a fix PR that adresses above comments too.

I can do that, but I need a Way to reproduce the Bug.

> I think I have reported that I cannot remove the pinned issue?

I know, this was 2 weeks ago. I wrote that I can't reproduce this and was asking for more Details to reproduce. Because I got no Response, I though this was maybe caused by a Database corruption or something else and the Problem was gone.

JakobDev commented on modules/structs/pull.go:
```diff
@@ -48,6 +48,8 @@ type PullRequest struct {
 	// swagger:strfmt date-time
 	Closed *time.Time `json:"closed_at"`
+
+	PinOrder int `json:"pin_order"`
```
Yes, you can pin Pull Requests

silverwind commented:
> I'm sorry, but I can reproduce this on a brand new setup on Firefox and Chrome. I can unpin Issues using the X without problems. Can you maybe investigate what causes the wrong ID in your case?

Just to confirm, this ID is meant to be the same as the issue ID seen in the URL, right?

JakobDev commented:
Thanks to your question I was able to found the Problem: The `ID` is the ID of the Issue, but I'm accidental using `issues_model.GetIssueByIndex` to get the Issue. As I tested this mostly in the First Repo that I have created with the same Issues, the ID was the same as the Index, so I didn't noticed it.
